src/codegen/aggregation.cpp

```diff
@@ -234,180 +361,335 @@ void Aggregation::DoAdvanceValue(CodeGen &codegen, llvm::Value *space,
       LOG_ERROR("%s", message.c_str());
       throw Exception{EXCEPTION_TYPE_UNKNOWN_TYPE, message};
     }
   }
 
   // Store the updated value in the appropriate slot
   PL_ASSERT(next.GetType().type_id != peloton::type::TypeId::INVALID);
-  storage_.SetValueSkipNull(codegen, space, agg_info.storage_index, next);
+  storage_.SetValueSkipNull(codegen, space, storage_index, next);
+}
+
+void Aggregation::DoNullCheck(CodeGen &codegen, llvm::Value *space,
+                              ExpressionType type, uint32_t storage_index,
+                              const codegen::Value &update,
+                              UpdateableStorage::NullBitmap &null_bitmap,
+                              llvm::Value *curr_val) const {
+  // This aggregate is NULL-able, we need to check if the update value is
+  // NULL, and whether the current value of the aggregate is NULL.
+
+  // There are two cases we handle:
+  // (1). If neither the update value or the current aggregate value are
+  //      NULL, we simple do the regular aggregation without NULL checking.
+  // (2). If the update value is not NULL, but the current aggregate **is**
+  //      NULL, then we just store this update value as if we're creating it
+  //      for the first time.
+  //
+  // If either the update value or the current aggregate are NULL, we have
+  // nothing to do.
+
+  llvm::Value *update_not_null = update.IsNotNull(codegen);
+  llvm::Value *agg_null = null_bitmap.IsNull(codegen, storage_index);
+
+  if (curr_val == nullptr) {
+    curr_val = null_bitmap.ByteFor(codegen, storage_index);
+  }
+
+  lang::If valid_update{codegen, update_not_null, "Agg.IfValidUpdate"};
+  {
+    lang::If agg_is_null{codegen, agg_null, "Agg.IfAggIsNull"};
+    {
+      // (2)
+      switch (type) {
+        case ExpressionType::AGGREGATE_SUM:
+        case ExpressionType::AGGREGATE_MIN:
+        case ExpressionType::AGGREGATE_MAX: {
+          storage_.SetValue(codegen, space, storage_index, update, null_bitmap);
+          break;
+        }
+        case ExpressionType::AGGREGATE_COUNT: {
+          codegen::Value one{type::BigInt::Instance(), codegen.Const64(1)};
+          storage_.SetValue(codegen, space, storage_index, one, null_bitmap);
+          break;
+        }
+        default: { break; }
+      }
+    }
+    agg_is_null.ElseBlock("Agg.IfAggIsNotNull");
+    {
+      // (1)
+      DoAdvanceValue(codegen, space, type, storage_index, update);
+    }
+    agg_is_null.EndIf();
+
+    // Merge the null value
+    null_bitmap.MergeValues(agg_is_null, curr_val);
+  }
+  valid_update.EndIf();
+
+  // Merge the null value
+  null_bitmap.MergeValues(valid_update, curr_val);
 }
 
 // Advance each of the aggregates stored in the provided storage space
 void Aggregation::AdvanceValues(
     CodeGen &codegen, llvm::Value *space,
-    const std::vector<codegen::Value> &next_vals) const {
+    const std::vector<codegen::Value> &next_vals,
+    const std::vector<codegen::Value> &grouping_keys) const {
```

Comment:
This method is too complex in its current form. Refactor by extracting the core aggregation logic into a separate function:
```
void Aggregation::AdvanceValue(CodeGen &codegen, llvm::Value *space, 
                               const std::vector<codegen::Value> &next_vals, 
                               const Aggregation::AggregateInfo &agg);
```

Then, `AdvanceValues` has logic to check distinct values, if necessary, and delegating to `AdvanceValue` to deal with individual aggregates. `AdvanceValue` then performs NULL-checking logic, delegating to `DoAdvanceValue` to physically perform the update.

Reply:
I had to extend the signature of the function to this:
```
void AdvanceValue(CodeGen &codegen, llvm::Value *space,
                                 const std::vector<codegen::Value> &next_vals,
                                 const Aggregation::AggregateInfo &agg,
                                 UpdateableStorage::NullBitmap &null_bitmap,
                                 llvm::Value *curr_val = nullptr) const;
```

When calling this function from inside the `lang::If` of the distinct check, we have to hand it the `NullBitmap`, as well as the `llvm::Value*` to be able to PHI merge this value after the `lang::If`. (same as in `DoNullCheck`)

Reply:
Just spoke with @tcm-marcel, I don't think passing along the `curr_val` is necessary.

Reply:
This is true, I changed it in https://github.com/cmu-db/peloton/pull/867/commits/2fbaeacff62889daa5f6c65e1916513a9ff72b63 and named it `null_byte_snapshot` to clarify its purpose.